### ISSUE TITLE
guest-fire.sh guest-water.sh and csf.deny

### ISSUE THREAD

**Issue by zanami** [omega8cc/boa#1438]
My server is under low volume ssh brute force attack. About 100=200 addresses get banned by lfd each hour.

BOA 4.0.2-dev

The problem is, my csf.deny rarely contains more than 50-60 addresses.

1.
If I'm correct, it's flushed each time [guest-fire.sh finds more than 6 copies of itself](https://github.com/omega8cc/boa/blob/master/aegir/tools/system/guest-fire.sh#L20) running and kills them.

gust-fire.sh fires each minute via /etc/crontab and tries to temp ban for 1 hour some IPs from /var/xdrago/monitor/ssh.log, but first it checks if the address is already banned (or allowed).

Then it waits for 5 seconds and does the same again (12 times total).

I'm not sure what's the reason for 6+ guest-fire.sh processes to exist. Maybe csf isn't responsive sometimes (it has to deal with lfd also) but then again only 1 guest-fire guest_guard() is actually working.

I had to comment out csf -df line of guest-guard.sh to keep my csf.deny

Can I schedule it to run every 5 minutes instead of 1 or something?

2.
guest-water.sh runs once per hour and it also clears csf.deny. Maybe not completely but...
I don't understand the /var/xdrago/monitor part and how it's connected with csf and especially lfd.

I wan't to keep those bastards banned until csf reaches its' DENY_IP_LIMIT and now all I have is a useless list of 20-80 addresses, so they can come back again ang again (and they do).

**Comment by omega8cc** [comment]
The guest-water.sh cleanup should run only once daily, not hourly, that was the old default, no longer needed. Now fixed. The reason guest-fire.sh kills too many copies of csf running is that it should never happen, but when it happens it's a sign of locked firewall, so full cleanup is the best way to handle that exception to avoid extended downtime.

**Comment by omega8cc** [comment]
The /etc/csf/csf.deny is managed by guest-water.sh not by guest-fire.sh and now will be updated/rewritten only once daily, as it should.

**Comment by zanami** [comment]
> The reason guest-fire.sh kills too many copies of csf running is that it should never happen

Not csf, guest-fire.sh itself, according to logs in  /var/log/fire-count.kill.log

> Please note that temporary blocks are not written in the /etc/csf/csf.deny

lfd reports permanent blocks and I can actually see them in csf.deny

> The /etc/csf/csf.deny is managed by guest-water.sh not by guest-fire.sh

Yet there is csf -df in guest-fire which flushes csf.deny if I’m not wrong. 

Thank you anyway. I’ll adjust cron job for guest-water and see if it helps.

**Comment by omega8cc** [comment]
> Not csf, guest-fire.sh itself, according to logs in /var/log/fire-count.kill.log

Doesn't matter, it's the same problem.

> lfd reports permanent blocks and I can actually see them in csf.deny

Permanent yes, temporary no.

> Yet there is csf -df in guest-fire which flushes csf.deny if I’m not wrong.

Correct, that was introduced as a safeguard after we have seen systems blocked or overloaded by too many csf related processes or too long lists of blocked IPs, which resulted with kind of self-DoS.

> I’ll adjust cron job for guest-water and see if it helps.

This will not help, this cron entry is auto-updated every 8 minutes..

> I noticed recent changes to both fire and water scripts. I guess it’s time to update BOA.

These changes are automatically applied as hot fixes unless you have added `_SKYNET_MODE=OFF`

**Comment by zanami** [comment]
> These changes are automatically applied as hot fixes unless you have added _SKYNET_MODE=OFF

I didn't have _SKYNET_MODE=OFF in .barracuda.cnf (no _SKYNET_MODE at all)
but no changes in /var/xdrago since manual barracuda update on October 14

What script is responsible for the hot fixes?

**Comment by omega8cc** [comment]
It's the running every 8 minutes `/var/xdrago/clear.sh` which downloads and runs the latest copy of https://github.com/omega8cc/boa/blob/master/BOA.sh.txt

You could try to run `bash /var/xdrago/clear.sh` manually, but please follow the guidelines so we could determine if there is any potential issue, like disk almost full, etc: https://github.com/omega8cc/boa/blob/master/CONTRIBUTING.txt#L13

**Comment by zanami** [comment]
Attackers were lazy last night but guest-fire.sh still killed itself (and csf) several times. I've added ps auxf to it to see what's going on.

[guest-fire.sh processes killed + ps auxf](https://gist.github.com/zanami/6d2222b24f54590ae6ada6c974e06652)

I only can tell that _ps aux | grep -v "grep" | grep --count "fire.sh"_ counts each guest-fire.sh as 2 because CRON uses /bin/dash  to run bash /var/xdrago/guest-fire.sh

Not sure if it's intended. I've added _grep -v "dash"_ but there's no activity recently so I can't tell if it helps.

[boa info](https://gist.github.com/zanami/79b246102cc4e478e216bdf60f69f15f)

**Comment by omega8cc** [comment]
All BOA scripts use bash, not dash, so not sure why are you getting these results?

Can you check the two log files for recent entries:

`/var/log/csf-count.kill.log`
`/var/log/fire-count.kill.log`

**Comment by omega8cc** [comment]
Ah, Ok, I can see the dash entries too.

**Comment by omega8cc** [comment]
Hope it's counted properly now.

**Comment by omega8cc** [comment]
That said, please make sure you can run without issues `bash /var/xdrago/clear.sh` and that it updates your scripts in `/var/xdrago`, if not, please send your server IP via omega8.cc/contact so we could check the communication from our end.

**Comment by zanami** [comment]
This happened once
> Wed Oct 23 21:25:01 +10 2019 Too many 6 guest-fire.sh processes killed
with my modifications of guest-fire (|grep -v "dash")

bash /var/xdrago/clear.sh runs fine but it does not update /var/xdrago

> # bash /var/xdrago/clear.sh
> 
> Cron Cleanup Completed

running BOA.sh.txt.hourly manually

> # bash BOA.sh.txt.hourly
> deny failed: 185.206.225.30 is in already in the deny file /etc/csf/csf.deny 1 times
> deny failed: 185.253.97.238 is in already in the deny file /etc/csf/csf.deny 1 times
> Redis is fine
> STATUS: Skynet Agent is Active, OK!
> STATUS: You can add the _SKYNET_MODE=OFF line in
> STATUS: /root/.barracuda.cnf to disable me, if needed.
> 
> BOA Meta Installer setup completed
> Please check INSTALL.txt and UPGRADE.txt at http://bit.ly/boa-docs
> Bye

It never calls update_agents because of these if statements

>     _VERSIONS_TEST=$(cat /var/log/barracuda_log.txt 2>&1)
>     if [[ "${_VERSIONS_TEST}" =~ "${_X_VERSION}" ]]; then
>       _VERSIONS_TEST_RESULT=OK
>       if [ -e "/root/.mstr.clstr.cnf" ] \
>         || [ -e "/root/.wbhd.clstr.cnf" ] \
>         || [ -e "/root/.dbhd.clstr.cnf" ] \
>         || [ -e "/root/.host8.cnf" ] \
>         || [ ! -e "/var/xdrago/conf/fpm-pool-foo-multi.conf" ] \
>         || [ "${_BENG_VS}" = "YES" ]; then
>         update_agents
>       fi

The first "if" passes and the second does not. There is fpm-pool-foo-multi.conf, no .cnf files, _BENG_VS in my case is obviously "NO".

It's set here in setup()

>   if [[ "${_CHECK_HOST}" =~ ".host8." ]] \
>     || [[ "${_CHECK_HOST}" =~ ".boa.io" ]] \
>     || [ "${_VMFAMILY}" = "VS" ]; then
>     _BENG_VS=YES
>   fi

VS is VServer and I'm using QEMU/KVM

**Comment by omega8cc** [comment]
But there’s second call to the update right below the else which should work.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by zanami** [comment]
The second call is inside "else" for the first (outer) "if" that checks for version (and that check returns true). So the second call is also unreachable.

**Comment by zanami** [comment]
>     if [[ "${_VERSIONS_TEST}" =~ "${_X_VERSION}" ]]; then
>       # we are here...............................................................
>       _VERSIONS_TEST_RESULT=OK
>       if [ -e "/root/.mstr.clstr.cnf" ] \
>         || [ -e "/root/.wbhd.clstr.cnf" ] \
>         || [ -e "/root/.dbhd.clstr.cnf" ] \
>         || [ -e "/root/.host8.cnf" ] \
>         || [ ! -e "/var/xdrago/conf/fpm-pool-foo-multi.conf" ] \
>         || [ "${_BENG_VS}" = "YES" ]; then
>       # ... but not here..........................................................
>         update_agents
>       fi
>     else
>       #... and not here...........................................................
>       update_agents
>       # force_rebuild
>       if [[ "${_VERSIONS_TEST}" =~ "BOA-3." ]] \
>         || [[ "${_VERSIONS_TEST}" =~ "BOA-2.4." ]] \
>         || [[ "${_VERSIONS_TEST}" =~ "BOA-2.3.8" ]]; then
>         _DO_NOTHING=YES
>       else
>         _DO_NOTHING=YES
>       fi
>     fi

**Comment by omega8cc** [comment]
Ah, right.. too many legacy leftovers here, now fixed. Thank you for pointing out the actual culprit.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by omega8cc** [comment]
It should work in the next 10 minutes.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by zanami** [comment]
clear.sh seems to work fine now, thanks
